### PR TITLE
feat(eks): bottlerocket versoin follows the cluster k8s versoin

### PR DESCRIPTION
**Pull request by pahud**
This PR will implicitly select correct bottlerocket variant in consist with the  cluster k8s version. For example, if the cluster k8s version is `1.17`, previously only bottlerocket `aws-k8s-1.15` variant is available. Now the variant will be `aws-k8s-1.17`

BREAKING CHANGE: Clusters previously running k8s version other than `1.15` and bottlerocket AMI(`aws-k8s-1.15` variant) will trigger AMI and node replacement.

Closes: #10188 

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by pahud** [comment]
Hi @iliapolo 

I am ready for the first round.

File: packages/@aws-cdk/aws-eks/lib/cluster.ts [hunk 1]
```diff
@@ -1484,20 +1486,29 @@ export class EksOptimizedImage implements ec2.IMachineImage {
...
+   */
+  readonly kubernetesVersion: string;
+}
+
 /**
  * Construct an Bottlerocket image from the latest AMI published in SSM
  */
-class BottleRocketImage implements ec2.IMachineImage {
-  private readonly kubernetesVersion?: string;
+export class BottleRocketImage implements ec2.IMachineImage {
```
Comment (iliapolo) [inline packages/@aws-cdk/aws-eks/lib/cluster.ts:1502]: Why did you decide to `export` this class now? I don't see it being passed anywhere

**Review by iliapolo — COMMENTED** [review]
I would classify this as a feature. The changelog should mention this behavior change.

File: packages/@aws-cdk/aws-eks/lib/cluster.ts (hunk 1 above)
Comment (pahud) [reply packages/@aws-cdk/aws-eks/lib/cluster.ts:1502]: Hi @iliapolo 

Looks like if I need unit the `BottleRocketImage()` in test.cluster.ts, I will need to export this class?

Is there any way I can unit test this class without exporting it?

圖片

thanks.

Comment (iliapolo) [reply packages/@aws-cdk/aws-eks/lib/cluster.ts:1502]: In general you would export it in a different file and not add that file to `index.ts`. In this case however I think the BottleRocketImage` is small enough to test it via the cluster itself.

Comment (pahud) [reply packages/@aws-cdk/aws-eks/lib/cluster.ts]: > @pahud In general you would export it in a different file and not add that file to `index.ts`. In this case however I think the BottleRocketImage` is small enough to test it via the cluster itself.

Thanks! It's in a seperate `bottlerocket.ts` now.

**Comment by pahud** [comment]
Hi @iliapolo 

Please take a look again. Thanks.

File: packages/@aws-cdk/aws-eks/lib/bottlerocket.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
+import * as ec2 from '@aws-cdk/aws-ec2';
```
Comment (iliapolo) [inline packages/@aws-cdk/aws-eks/lib/bottlerocket.ts:1]: put this in a `private` directory so it's clear. This is our normal convention.

**Comment by iliapolo** [comment]
@pahud 

> Clusters previously running k8s version other than 1.15 and bottlerocket AMI(aws-k8s-1.15 variant) will trigger AMI and node replacement.

I'm not sure how I feel about this. While it's as intrusive as cluster replacement, it still feels a bit off.

@eladb What do you think? Maybe worth waiting on this feature or implementing it in a non breaking way? i.e opt-in for bottle rocket version tracking.

File: packages/@aws-cdk/aws-eks/lib/private/bottlerocket.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+  /**
+   * Return the correct image
+   */
+  public getImage(scope: Construct): ec2.MachineImageConfig {
+    const ami = ssm.StringParameter.valueForStringParameter(scope, this.amiParameterName);
+    return {
+      imageId: ami,
+      osType: ec2.OperatingSystemType.LINUX,
+      userData: ec2.UserData.custom(''),
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks/lib/private/bottlerocket.ts:41]: Why is this required?

**Review by eladb — APPROVED** [review]

**Comment by pahud** [comment]
Hi @eladb @iliapolo 

Just fixed the build error.

File: packages/@aws-cdk/aws-eks/lib/private/bottlerocket.ts (hunk 3 above)
Comment (pahud) [reply packages/@aws-cdk/aws-eks/lib/private/bottlerocket.ts:41]: > Why is this required?

We had exactly the same discussion here
https://github.com/aws/aws-cdk/pull/7282#discussion_r407020235

**Comment by pahud** [comment]
Hi @eladb @iliapolo 

Anything else we should add or fix for this PR?

File: packages/@aws-cdk/aws-eks/lib/private/bottlerocket.ts (hunk 3 above)
Comment (eladb) [reply packages/@aws-cdk/aws-eks/lib/private/bottlerocket.ts:41]: Can we chance it to optional? (let's at least open an issue?)

**Review by eladb — APPROVED** [review]
